File: scripts/algorithms/V/Verbal Arithmetic Puzzle/Verbal Arithmetic Puzzle.java
// Runtime: 7 ms (Top 96.2%) | Memory: 40.32 MB (Top 56.2%)

class Solution {
     public static boolean isSolvable(String[] words, String result) {
	    // reverse all strings to facilitate add calculation.
        for (int i = 0; i < words.length; i++) {
            words[i] = new StringBuilder(words[i]).reverse().toString();
        }
        result = new StringBuilder(result).reverse().toString();
        if (!checkLength(words, result)) {
            return false;
        }
        boolean[] visited = new boolean[10];    // digit 0, 1, ..., 9
        int[] chToDigit = new int[26];
        Arrays.fill(chToDigit, -1);
        return dfs(0, 0, 0, visited, chToDigit, words, result);
    }

    /**
     * Elminate the case where result is too long
     * word1:  AAA
     * word2:  BBB
     * result: XXXXXXXXX
     */
    private static boolean checkLength(String[] words, String result) {
        int maxLen = 0;
        for (String word : words) {
            maxLen = Math.max(maxLen, word.length());
        }
        return result.length() == maxLen || result.length() == maxLen + 1;
    }

    /*
    Put all words like this:
    w1:     ABC
    w2:     EF
    w3:     GHIJ
    result: KLMNO
    i, is the row
    j,  is the column
    carrier, the one contributed from previous calculation
    chToDigit, 26 int array, which records choosen digit for 'A', 'B', 'C', ... If not choosen any, default is -1 
     */
    private static boolean dfs(int i, int j, int carrier, boolean[] visited, int[] chToDigit, String[] words, String result) {
        if (i == words.length) {
            char ch = result.charAt(j);
            // (i, i) at bottom right corner. final check
            if (j == result.length() - 1) {
                // 1. check if carrier is equal or greater than 10. If so, false.
                if (carrier >= 10) {
                    return false;
                }
                // 2. check if result.length() > 1 && result.lastCh is zero. If so the false.
                if (j > 0 && j == result.length() - 1 && chToDigit[ch - 'A'] == 0) {
                    return false;
                }
                // not selected, can select any. True.
                if (chToDigit[ch - 'A'] == -1) {
                    System.out.println(Arrays.toString(chToDigit));
                    return true;
                } else {    // if selected, check if it matches with carrier. Also, carrier can't be 0. result = '00' is invalid
                    return chToDigit[ch - 'A'] == carrier;
                }
            } else {    // reached normal result line.
                // 1. if not selected. Use current carrier's unit digit
                if (chToDigit[ch - 'A'] == -1) {
                    int selectedDigit = carrier % 10;
                    // For example carrier = 13. selectedDigit = 3. ch = 'H'. Should set 3 to 'H'.
                    // But 3 is already taken by 'B' previously. So wrong.
                    if (visited[selectedDigit]) {
                        return false;
                    }
                    visited[selectedDigit] = true;
                    chToDigit[ch - 'A'] = selectedDigit;
                    if (dfs(0, j + 1, carrier / 10, visited, chToDigit, words, result)) {
                        return true;
                    }
                    chToDigit[ch - 'A'] = -1;
                    visited[selectedDigit] = false;
                } else { // 2. selected
                    // just need to check if ch.digit equals to unit digit.
                    if (chToDigit[ch - 'A'] != carrier % 10) {
                        return false;
                    }
                    boolean ans = dfs(0, j + 1, carrier / 10, visited, chToDigit, words, result);
                    return ans;
                }
            } //
        } else { // normal word
            String word = words[i];
            // 1. check if j is equal or greater than word.len. If so pass to next word.
            if (j >= word.length()) {
                boolean ans = dfs(i + 1, j, carrier, visited, chToDigit, words, result);
                return ans;
            }
            // 2. check if it's last ch, word.len is greater than 1, and is '0'. If so false;
            if (j == word.length() - 1 && word.length() > 1 && chToDigit[word.charAt(j) - 'A'] == 0) {
                return false;
            }
            char ch = word.charAt(j);
            // 3. check if word.ch is selected. Just add current digit and move to next word.
            if (chToDigit[ch - 'A'] != -1) {
                int newSum = carrier + chToDigit[ch - 'A'];
                boolean ans = dfs(i + 1, j, newSum, visited, chToDigit, words, result);
                return ans;
            } else {
                for (int k = 0; k < visited.length; k++) {
                    if (visited[k]) {
                        continue;
                    }
                    visited[k] = true;
                    chToDigit[ch - 'A'] = k;
                    int newSum = k + carrier;
                    boolean ans = dfs(i + 1, j, newSum, visited, chToDigit, words, result);
                    if (ans) {
                        return true;
                    }
                    visited[k] = false;
                    chToDigit[ch - 'A'] = -1;
                }
            }
        }
        return false;
    }
}

File: scripts/algorithms/V/Verbal Arithmetic Puzzle/Verbal Arithmetic Puzzle.cpp
class Solution {
    int limit = 0;
    unordered_map<char, int> c2i;
    unordered_map<int, char> i2c;
    bool helper(vector<string> &words, string &result, int digit, int wid, int sum)
    {
        if(digit==limit)
            return sum == 0;
        if(wid==words.size())
        {
            if(c2i.count(result[digit])==0 && i2c.count(sum%10)==0)
            {
                if(sum%10==0 && digit+1 == limit)
                    return false; //Means leading zeros
                c2i[result[digit]] = sum%10;
                i2c[sum%10] = result[digit];
                bool tmp = helper(words, result, digit+1, 0, sum/10); //0 because it will
                //go one digit at a time until all digits in words are finished
                c2i.erase(result[digit]);
                i2c.erase(sum%10);
                return tmp;
            }
            else if(c2i.count(result[digit]) && c2i[result[digit]] == sum%10) //c2i for that exits and is equal to sum%10, remember default will be 0 and sum%10 can also be 0
            {
                if(digit+1 == limit && c2i[result[digit]]==0)
                    return false; //again same condition of leading zeros
                return helper(words, result, digit+1, 0, sum/10);
            }
            else //if result[digit] exists in c2i but is not equal to sum%10
                return false;
        }
        if(digit>=words[wid].length()) //digit>current words length
            return helper(words, result, digit, wid+1, sum); //go to next word at same position
        //(digit) and sum, also going wid + 1 that is why checking wid limit in last condition already
        if(c2i.count(words[wid][digit]))
        {
            if(digit+1 == words[wid].length() && words[wid].length() > 1 && c2i[words[wid][digit]]==0)
                return false; //here we checking if there is no leading 0 in the word itself
            return helper(words, result, digit, wid+1, sum+c2i[words[wid][digit]]);
        }
        for(int i=0; i<10; i++)
        {
            if(digit+1==words[wid].length() && i==0 && words[wid].length()>1)
                continue;
            if (i2c.count(i))
                continue;
            c2i[words[wid][digit]] = i;
            i2c[i] = words[wid][digit];
            bool tmp = helper(words, result, digit, wid+1, sum+i);
            c2i.erase(words[wid][digit]);
            i2c.erase(i);
            if(tmp)
                return true;
        }
        return false;
    }
    
public:
    bool isSolvable(vector<string>& words, string result) {
        limit = result.length();
        for(auto s:words)
        {
            if(s.length()>limit)
                return false;
        }
        for(auto &s:words)
        {
            reverse(s.begin(), s.end());
        }
        reverse(result.begin(), result.end());
        return helper(words, result, 0, 0, 0);
    }
};